tests/test_eigen_tensor.py

```diff
@@ -0,0 +1,135 @@
+import pytest
+
+np = pytest.importorskip("numpy")
+m = pytest.importorskip("pybind11_tests.eigen_tensor")
+
+tensor_ref = np.array(
+    [
+        [[0.0, 15.0], [3.0, 18.0], [6.0, 21.0], [9.0, 24.0], [12.0, 27.0]],
+        [[1.0, 16.0], [4.0, 19.0], [7.0, 22.0], [10.0, 25.0], [13.0, 28.0]],
+        [[2.0, 17.0], [5.0, 20.0], [8.0, 23.0], [11.0, 26.0], [14.0, 29.0]],
+    ],
+)
+
+indices = (2, 3, 1)
+
+
+def assert_equal_tensor_ref(mat, writeable=True, modified=0):
+    assert mat.flags.writeable == writeable
+
+    copy = np.array(tensor_ref)
+    if modified != 0:
+        copy[indices] = modified
+
+    np.testing.assert_array_equal(mat, copy)
+
+
+def test_convert_tensor_to_py():
+    assert_equal_tensor_ref(m.copy_tensor())
+    assert_equal_tensor_ref(m.copy_fixed_tensor())
+    assert_equal_tensor_ref(m.copy_const_tensor())
+
+    assert_equal_tensor_ref(m.move_tensor())
+    assert_equal_tensor_ref(m.move_fixed_tensor())
+
+    assert_equal_tensor_ref(m.take_tensor())
+    assert_equal_tensor_ref(m.take_fixed_tensor())
+
+    assert_equal_tensor_ref(m.reference_tensor())
+    assert_equal_tensor_ref(m.reference_tensor_v2())
+    assert_equal_tensor_ref(m.reference_fixed_tensor())
+
+    assert_equal_tensor_ref(m.reference_view_of_tensor())
+    assert_equal_tensor_ref(m.reference_view_of_fixed_tensor())
+    assert_equal_tensor_ref(m.reference_const_tensor(), writeable=False)
+    assert_equal_tensor_ref(m.reference_const_tensor_v2(), writeable=False)
+
+
+def test_bad_cpp_to_python_casts():
+    with pytest.raises(Exception):
```

Comment:
Wait, do we really return BaseExceptions? Pretty sure these are RuntimeException/TypeException at the very least?

Reply:
These are RuntimeExceptions as they indicate bugs in the C++ code, not the Python code.

Reply:
I'd definitely try to narrow this down, usually it's really easy and I often ~cheat like this: copy the inner line up with unindent, run the test, copy-paste the exception type from the error message here. (At that point I usually ask myself it the exception type makes sense, to not cheat too badly.)